rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,546 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <chrono>
+#include <string>
+#include <thread>
+
+#include "rcl/rcl.h"
+#include "rcl/subscription.h"
+#include "rcl/error_handling.h"
+
+#include "test_msgs/msg/primitives.h"
+#include "rosidl_generator_c/string_functions.h"
+
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+
+using std::chrono::milliseconds;
+using std::chrono::seconds;
+
+constexpr seconds LIVELINESS_LEASE_DURATION_IN_S(1);
+constexpr seconds DEADLINE_PERIOD_IN_S(1);
+
+#ifdef RMW_IMPLEMENTATION
+# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME(NAME, SUFFIX) CLASSNAME_(NAME, SUFFIX)
+#else
+# define CLASSNAME(NAME, SUFFIX) NAME
+#endif
+
+class CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION) : public ::testing::Test
+{
+public:
+  void SetUp()
+  {
+    is_opensplice =
+      std::string(rmw_get_implementation_identifier()).find("rmw_opensplice") == 0;
+    is_fastrtps =
+      std::string(rmw_get_implementation_identifier()).find("rmw_fastrtps") == 0;
+
+    is_unsupported = is_fastrtps;
```

Comment:
Alternatively, you could determine this programmatically, e.g. call one of the functions and see if it returns unsupported. That way this could be enabled/disabled automatically even for new rmw's. On the other hand, they should all support this feature as soon as possible, as we already discussed on the design document, I don't think optional features are a good thing to have long term.

Reply:
Will add the `TODO`.

Many of these tests are testing against `rcl_publisher_init()`/`rcl_subscription_init()`, which we left untouched since the changes made for the QoS features are already quite a sizeable scope, and `rcl_publisher_init()`/`rcl_subscription_init()` don't seem to have the return unsupported behavior.

I agree that having optional features is not a good thing. It'll be up to FastRTPS to provide the features.

Reply:
Ok, sounds good.